# 01. Import the required libraries

In [2]:
import pandas as pd
import numpy as np
import os

# 02. Importing the data

In [59]:
# Importing departments.csv
df_dep = pd.read_csv(r'/Users/omkargopalakrishnan/Instacart Basket Analysis/Original Data/departments.csv', index_col = False)

In [60]:
# Importing orders.csv
df_ords = pd.read_csv(r'/Users/omkargopalakrishnan/Instacart Basket Analysis/Original Data/orders.csv', index_col = False)

In [72]:
# Import products.csv
df_prods = pd.read_csv(r'/Users/omkargopalakrishnan/Instacart Basket Analysis/Original Data/products.csv', index_col = False)

In [4]:
# Telling Python to remember a main folder path
path = r'/Users/omkargopalakrishnan/Instacart Basket Analysis/'

In [5]:
path

'/Users/omkargopalakrishnan/Instacart Basket Analysis/'

In [21]:
# Simplify the import function
df_dep = pd.read_csv(os.path.join(path, 'Original Data', 'departments.csv'), index_col = False)

# Data Wrangling Procedures

## Dropping Columns

In [61]:
df_ords.drop(columns = ['eval_set']) # drop the column "eval_set" from tne “orders.csv” dataframe

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [62]:
df_ords = df_ords.drop(columns = ['eval_set']) # remove the "eval_set" column from df_ords and have the result of that removal be the new df_ords

In [64]:
df_ords ['days_since_prior_order'].value_counts(dropna = False) # looking for missing values(NaN) using dropna = False 

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

## Renaming Columns

In [65]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True) # renaming a column

In [66]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

## Changing a Variable’s Data Type

In [67]:
df_ords['order_id'] = df_ords['order_id'].astype('str') # change the data type of the "order_id" column from a number to a string

In [69]:
df_ords ['order_id'].dtype # return the data type of your new "order_id" column. You should get an “O,” which stands for object, pandas’ version of the string

dtype('O')

## Transposing Data

In [22]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [30]:
df_dep.T #table format transposed, switching the rows and columns

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [32]:
df_dep_t = df_dep.T # creating a new dataframe from the transposed version and call it df_dep_t

In [35]:
df_dep_t #check out your new dataframe by typing its name

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [36]:
# adding an index to the dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [37]:
# After creating an index column in the previous step, we can create a new header for the dataframe. 
# This will be done in three steps: 
# 1) Create a new header; 
# 2) Remove the first row in the dataframe; and 
# 3) Add the new header. This all sounds a little abstract, so let’s walk through the process together.
new_header = df_dep_t.iloc[0]

In [38]:
new_header # see the contents of your first row

0    department
Name: department_id, dtype: object

In [40]:
df_dep_t_new = df_dep_t[1:] # Now that you’ve made a copy of your first row, you need to delete it. Otherwise, when you add the first row back in as your header, you’ll have two of them: one as a header and one as a row. That’s not what you want! As such, let’s create a new dataframe that only copies over rows beyond the first row.

In [41]:
df_dep_t_new # The resulting df_dep_t_new dataframe will look exactly like the old dataframe only without the first row.

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [42]:
df_dep_t_new.columns = new_header # set your new_header variable as the new header for your df_dep_t_new dataframe

In [43]:
df_dep_t_new # Proper column headers!

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Data Dictionaries

In [44]:
data_dict = df_dep_t_new.to_dict('index') # create a data dictionary that contains meanings for the values in the "department_id" column

In [15]:
# Import the “products.csv” file into Jupyter
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)

In [45]:
data_dict # calling the new data_dict variable

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [46]:
# check the df_prods dataframe
df_prods.head

<bound method NDFrame.head of        product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104        

In [47]:
# Using the data dictionary to determine what department_id 19 stands for
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting

In [50]:
df_snacks =  df_prods[df_prods['department_id']==19] # Create a subset for the df_prods dataframe that contains only data from the snacks department

In [51]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [52]:
df_prods[df_prods['department_id']==19] #Enclosing df_prods['department_id']==19 inside df_prods[] tells Python to map the true/false procedure you witnessed in the previous output onto the columns within the df_prods dataframe—the result being a list of only those values within df_prods that are true (and, subsequently, only those values within df_prods that have a "department_id" of 19).

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [53]:
df_snacks =  df_prods[df_prods['department_id']==19] # save the results above into a new dataframe called df_snacks

In [54]:
df_snacks.head() # check out the dataframe

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [56]:
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19] # A second way of achieving the above same results which is a bit simpler. This involves use of the loc function. The loc function is similar to the iloc function you used previously, but while iloc looks for an integer position, loc looks for a column name. In your case, you want this function to look for instances where the value of "department_id" is 19 within the df_prods database

In [57]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])] # Another way to achive the same results. In this second version, rather than using an equals sign, you’re telling the loc function to look into a list: isin([19]). This is handy because you could use more than one value here if you wanted to check multiple "department_ids" at the same time (isin([17,18,19])).

# Task

### 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format

In [75]:
# Changing 'user_id' column to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [76]:
# Check the changed data type of the variable
df_ords['user_id'].dtype

dtype('O')

### 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe

In [77]:
# Check the dataframe 
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [79]:
# Changing the columns without overwriting the dataframe
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_previous_order'}, inplace = False)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_previous_order  
0                              NaN  
1                             15.0  
2                             21.0  
3                             29.0  
4                             28.0  
...                            ...  
3421078                       29.0  
3421079                       30.0  
3421080                       18.0  
3421081                        7.0  
3421082                       30.0  

[3421083 rows x 6 columns]

### 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings

In [80]:
df_ords['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

#### Answer: The busiest hour for placing orders is 10am

### 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary

In [81]:
# Checking the data dictionary for the department name for 4
print(data_dict.get('4'))

{'department': 'produce'}


#### Answer: The value 4 in the 'department_id' column represents the department "produce"

### 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information

In [98]:
# Checking the department number for the category 'breakfast'
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [83]:
# Create a df_breakfast dataframe showing the products with the 'department_id' 14
df_breakfast = df_prods[df_prods['department_id'] == 14]

In [97]:
# Checking the freshly created dataframe 
df_breakfast.head()

In [99]:
# Checking the frequency of each distinct row in the Dataframe
df_breakfast['department_id'].value_counts()

department_id
14    1116
Name: count, dtype: int64

### 7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [86]:
# Checking the data dictionary 
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [87]:
# Create a dataframe df_dinner_parties with products with the 'department_ids': 5(alcohol), 7(beverages), 12(meat seafood), 20(deli)
df_dinner_parties = df_prods[df_prods['department_id'].isin([5,7,12,20])]

In [88]:
df_dinner_parties.head(15)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   
43          44                        Sparkling Raspberry Seltzer       115   
48          49      Vegetarian Grain Meat Sausages Italian - 4 CT        14   
51          52                                Mirabelle Brut Rose       134   
52          53              Whole Leaf Pure Aloe With Lemon Juice        98   
61          62            Premium Deli Oven Roasted Turkey Breast        96   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5  
43              7    12.2  
48             20    10.1  
51              5    14.4  
52              7     4.7  
61             20    14.6

In [89]:
df_dinner_parties['department_id'].value_counts()

department_id
7     4365
20    1322
5     1056
12     907
Name: count, dtype: int64

### 8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [90]:
# Finding the number of rows in df_dinner_parties
df_dinner_parties.shape

(7650, 5)

#### Answer: 7650 rows

### 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.


In [92]:
# Create a df_user_1 dataframe with user_id of 1
df_user_1 = df_ords[df_ords['user_id'] == '1']

In [93]:
df_user_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

### 10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [94]:
df_user_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### 11. Check the organization and structure of your notebook. Be sure to include section headings and code comments

### Done

### 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder

In [95]:
# Exporting df_ords dataframe
df_ords.to_csv(os.path.join(path,'Prepared Data', 'orders_wrangled.csv'))

### 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format

In [96]:
# Exporting df_dep_t_new dataframe
df_dep_t_new.to_csv(os.path.join(path, 'Prepared Data', 'departments_wrangled.csv'))